src/libtls/tls_server.c

```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
 	bool curves_received;
 };
 
+/**
+ * Create a hashtable of an intersection of server and peer supported key types
+ */
+static hashtable_t *create_common_key_types(chunk_t hashsig,
+											tls_version_t version_min,
+											tls_version_t version_max)
+{
+	hashtable_t *ht;
+	enumerator_t *enumerator;
+	int *k;
+	key_type_t *v;
+	uint16_t current;
+
+	ht = tls_get_supported_key_types(version_min, version_max);
+	if (!ht)
+	{
+		return ht;
+	}
+	enumerator = ht->create_enumerator(ht);
+	while (enumerator->enumerate(enumerator, &k, &v))
+	{
+		bio_reader_t *reader;
+		bool found = FALSE;
+
+		reader = bio_reader_create(hashsig);
+		while (reader->remaining(reader) && reader->read_uint16(reader, &current))
+		{
+			key_type_t type = tls_signature_scheme_to_key_type(current);
+			if (*v == type)
+			{
+				found = TRUE;
+				break;
+			}
+		}
+		reader->destroy(reader);
+		if (!found)
+		{
+			ht->remove_at(ht, enumerator);
+			free(v);
+		}
+	}
+	enumerator->destroy(enumerator);
+	return ht;
+}
+
 /**
  * Find a cipher suite and a server key
  */
 static bool select_suite_and_key(private_tls_server_t *this,
 								 tls_cipher_suite_t *suites, int count)
 {
+	hashtable_t *key_types;
+	enumerator_t *enumerator;
+	tls_version_t version_min, version_max;
 	private_key_t *key;
-	key_type_t type;
+	key_type_t type = 0, *v;
+	int *k;
 
-	key = lib->credmgr->get_private(lib->credmgr, KEY_ANY, this->server,
-									this->server_auth);
+	version_min = this->tls->get_version_min(this->tls);
+	version_max = this->tls->get_version_max(this->tls);
+	key_types = create_common_key_types(this->hashsig, version_min, version_max);
+	if (!key_types)
+	{
+		DBG1(DBG_TLS, "no common signature algorithm found");
```

Comment:
`algorithm` -> `algorithms`. The number of elements should probably be checked as this always returns a valid pointer it seems.